libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -85,43 +147,57 @@ internal class RouteRefreshController(
+        newRoute: NavigationRoute,
         oldRoute: NavigationRoute,
         currentLegIndex: Int,
-        onRoutesRefreshed: (List<NavigationRoute>) -> Unit
-    ) = object : NavigationRouterRefreshCallback {
+    ) {
+        logI("Successful route refresh", LOG_CATEGORY)
+        val routeDiffs = routeDiffProvider.buildRouteDiffs(
+            oldRoute,
+            newRoute,
+            currentLegIndex,
+        )
+        if (routeDiffs.isEmpty()) {
+            logI("No changes to route annotations", LOG_CATEGORY)
+        } else {
+            for (diff in routeDiffs) {
+                logI(diff, LOG_CATEGORY)
+            }
+        }
+    }
 
-        override fun onRefreshReady(route: NavigationRoute) {
-            logI("Successful route refresh", LOG_CATEGORY)
-            val routeDiffs = routeDiffProvider.buildRouteDiffs(
-                oldRoute,
+    private suspend fun requestRouteRefresh(route: NavigationRoute, legIndex: Int) =
+        suspendCancellableCoroutine<RouteRefreshResult> { continuation ->
+            val requestId = directionsSession.requestRouteRefresh(
                 route,
-                currentLegIndex,
-            )
-            if (routeDiffs.isEmpty()) {
-                logI("No changes to route annotations", LOG_CATEGORY)
-            } else {
-                for (diff in routeDiffs) {
-                    logI(diff, LOG_CATEGORY)
+                legIndex,
+                object : NavigationRouterRefreshCallback {
+                    override fun onRefreshReady(route: NavigationRoute) {
+                        continuation.resume(RouteRefreshResult.Success(route))
+                    }
+
+                    override fun onFailure(error: NavigationRouterRefreshError) {
+                        continuation.resume(RouteRefreshResult.Fail(error))
+                    }
                 }
+            )
+            continuation.invokeOnCancellation {
+                directionsSession.cancelRouteRefreshRequest(requestId)
             }
-            val directionsSessionRoutes = directionsSession.routes.toMutableList()
-            if (directionsSessionRoutes.isNotEmpty()) {
-                directionsSessionRoutes[0] = route
-                onRoutesRefreshed(directionsSessionRoutes)
-            }
-            currentRequestId = null
         }
 
-        override fun onFailure(error: NavigationRouterRefreshError) {
-            logE(
-                "Route refresh error: ${error.message} throwable=${error.throwable}",
-                LOG_CATEGORY
-            )
-            currentRequestId = null
-        }
+    private sealed class RouteRefreshResult {
+        data class Success(val route: NavigationRoute) : RouteRefreshResult()
+        data class Fail(val error: NavigationRouterRefreshError) : RouteRefreshResult()
+    }
+
+    private suspend fun waitForever(message: String): List<NavigationRoute> {
+        logW("Refresh won't happen because $message", LOG_CATEGORY)
```

Comment:
```suggestion
        logI("Route refresh disabled. Reason: $message", LOG_CATEGORY)
```
Refresh being disabled is not necessarily a problem, I'm suggesting to decrease the log level.